datafusion/common/src/file_options/parquet_writer.rs

```diff
@@ -51,38 +58,53 @@ impl ParquetWriterOptions {
     }
 }
 
-impl TryFrom<&TableParquetOptions> for ParquetWriterOptions {
-    type Error = DataFusionError;
-
-    fn try_from(parquet_table_options: &TableParquetOptions) -> Result<Self> {
+impl TableParquetOptions {
+    #[deprecated(
+        since = "44.0.0",
+        note = "Please use `TableParquetOptions::into_writer_properties_builder` and `TableParquetOptions::into_writer_properties_builder_with_arrow_schema`"
+    )]
+    pub fn try_from(table_opts: &TableParquetOptions) -> Result<ParquetWriterOptions> {
         // ParquetWriterOptions will have defaults for the remaining fields (e.g. sorting_columns)
         Ok(ParquetWriterOptions {
-            writer_options: WriterPropertiesBuilder::try_from(parquet_table_options)?
-                .build(),
+            writer_options: table_opts.into_writer_properties_builder()?.build(),
         })
     }
-}
-
-impl TryFrom<&TableParquetOptions> for WriterPropertiesBuilder {
-    type Error = DataFusionError;
 
     /// Convert the session's [`TableParquetOptions`] into a single write action's [`WriterPropertiesBuilder`].
     ///
     /// The returned [`WriterPropertiesBuilder`] includes customizations applicable per column.
-    fn try_from(table_parquet_options: &TableParquetOptions) -> Result<Self> {
+    pub fn into_writer_properties_builder(&self) -> Result<WriterPropertiesBuilder> {
+        self.into_writer_properties_builder_with_arrow_schema(None)
+    }
+
+    /// Convert the session's [`TableParquetOptions`] into a single write action's [`WriterPropertiesBuilder`].
+    ///
+    /// The returned [`WriterPropertiesBuilder`] includes customizations applicable per column,
+    /// as well as the arrow schema encoded into the kv_meta at [`ARROW_SCHEMA_META_KEY`].
+    pub fn into_writer_properties_builder_with_arrow_schema(
```

Comment:
I think making setting the arrow schema a different function might be more consistent with the rest of this API https://docs.rs/datafusion/latest/datafusion/config/struct.ParquetOptions.html


So for example I would expect that `ParquetOptions::into_writer_properties_builder` would always set the arrow metadata to be consistent with how `WriterProperties` works 🤔 

If someone doesn't want the arrow metadata, I would expect an option to disable it, like https://docs.rs/parquet/53.3.0/parquet/arrow/arrow_writer/struct.ArrowWriterOptions.html#method.with_skip_arrow_metadata

So that might mean that `TableParquetOptions` has a field like `skip_arrow_metadata`

And then depending on the value of that field, `into_writer_properties_builder` would set/not set the metadata appropriately

That would also avoid having to change the conversion API

Reply:
> So for example I would expect that ParquetOptions::into_writer_properties_builder would always set the arrow metadata to be consistent with how WriterProperties works 🤔

The [ParquetOptions](https://docs.rs/datafusion/latest/datafusion/config/struct.ParquetOptions.html) does not have kv_metadata. That is held within the parent structure [TableParquetOptions](https://docs.rs/datafusion/latest/datafusion/common/config/struct.TableParquetOptions.html). That is why the existing [TryFrom code only constructs the writer properties from the TableParquetOptions](https://github.com/apache/datafusion/blob/a08dc0af8f798afe73a2fdf170ab36e72ad7e782/datafusion/common/src/file_options/parquet_writer.rs#L66). Not the ParquetOptions.

Reply:
Here is the change: https://github.com/apache/datafusion/pull/13866/commits/f2f9b00abd9e783fdbc32e8b167251423f483fbc

I also made another change to prevent the user errors I was worried about: https://github.com/apache/datafusion/pull/13866/commits/c5ad794bb9652e84e9ad7a66f15bef964fb2a4dc